### ISSUE TITLE
Detect unused namespaces

### ISSUE THREAD

**Issue by ivarref** [borkdude/clj-kondo#141]
A namespace you use/require could be removed.

Eastwood has [this feature](https://github.com/jonase/eastwood#unused-namespaces).

Edit: Thanks for a good and fast project!

**Comment by borkdude** [comment]
Thanks for using clj-kondo and reporting this issue.

**Comment by borkdude** [comment]
@ivarref I've got a rough first version ready. It lints only `:require` and only if it doesn't contain `:all`. I think that's the best I can do with static analysis without generating a lot of false positives.

![Screenshot 2019-05-07 23 41 18](https://user-images.githubusercontent.com/284934/57335054-cb4a1480-7121-11e9-9397-30acd31cbb87.png)

are you using the linux or MacOS binary? I can provide you a preview of this, so you can test if you want.

**Comment by ivarref** [comment]
Hi, and thanks for replying quickly!

I’m on linux and am happy to test.

**Comment by borkdude** [comment]
@ivarref You can try this one out:

https://2201-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.05.07-alpha-SNAPSHOT-linux-amd64.tar.gz

**Comment by borkdude** [comment]
TODO:
- [ ] recognize usage of namespace when using namespaced keywords

**Comment by borkdude** [comment]
@ivarref This version also support detecting unused namespaces when you use namespaced keywords:

https://2205-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.05.07-alpha-SNAPSHOT-linux-amd64.tar.gz

**Comment by borkdude** [comment]
TODO:

- [ ] fix false positive in below case
```
(ns foo (:require [clojure.set :as set :refer [difference]]))
    (reduce! difference #{} [])
```

**Comment by borkdude** [comment]
Updated version for linux:

https://2215-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.05.07-alpha-SNAPSHOT-linux-amd64.tar.gz

**Comment by ivarref** [comment]
Looks good to me!

One thing that hit me was the [namespaced maps](https://github.com/xsc/rewrite-clj/issues/54) issue which I see you are already working on (and which my code is full of).

Thanks and kind regards!

**Comment by borkdude** [comment]
@ivarref This version should do better on the namespaced maps:

https://2240-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.05.07-alpha-SNAPSHOT-linux-amd64.tar.gz

There are still a few details I want to look at, but meanwhile any false positives or bug reports would be appreciated.

**Comment by borkdude** [comment]
I merged to master now and this should end up in the next version.

https://2278-176829714-gh.circle-artifacts.com/0/release/clj-kondo-2019.05.07-alpha-SNAPSHOT-linux-amd64.tar.gz

**Comment by ivarref** [comment]
Thanks!

Best regards

**Comment by borkdude** [comment]
Released: https://github.com/borkdude/clj-kondo/releases/tag/v2019.05.12-alpha

**Comment by ProjectFrank** [comment]
It appears that `clj-kondo` does not recognize the usage of a namespace when using namespaced keywords in the destructuring expressions such as `{:keys [::namespaced/keyword]}`, at least in the context of destructuring of `fn` args.

**Comment by borkdude** [comment]
@ProjectFrank This is probably fixed in this issue: https://github.com/borkdude/clj-kondo/issues/164. Can you test and report back?

**Comment by ProjectFrank** [comment]
Yep, looks fixed!

**Comment by borkdude** [comment]
Thank you. I hope to release a new version this weekend.
